File: platform/core-api/src/com/intellij/ide/util/RunOnceUtil.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.ide.util;

import com.intellij.openapi.project.Project;
import com.intellij.util.PlatformUtils;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

/**
 * Run task once in the history for project or application.
 *
 * @author Konstantin Bulenkov
 */
public final class RunOnceUtil {
  /**
   * Perform the task if it was not performed before for the given project.
   *
   * @param id unique id for the task
   * @return {@code true} if task was performed, {@code false} if task had already been performed before.
   */
  public static boolean runOnceForProject(@NotNull Project project, @NotNull @NonNls String id, @NotNull Runnable task) {
    return _runOnce(PropertiesComponent.getInstance(project), id, false, task);
  }

  /**
   * Perform the task if it was not performed before for the given project.
   *
   * @param id unique id for the task
   * @param isIDEAware if {@code true}, the task will be performed for each IDE
   * @return {@code true} if task was performed, {@code false} if task had already been performed before.
   */
  public static boolean runOnceForProject(@NotNull Project project, @NotNull @NonNls String id, boolean isIDEAware, @NotNull Runnable task) {
    return _runOnce(PropertiesComponent.getInstance(project), id, isIDEAware, task);
  }

  /**
   * Perform the task if it was not performed before for this application.
   *
   * @param id unique id for the task
   * @return {@code true} if task was performed, {@code false} if task had already been performed before.
   */
  public static boolean runOnceForApp(@NotNull @NonNls String id, @NotNull Runnable task) {
    return _runOnce(PropertiesComponent.getInstance(), id, false, task);
  }

  private static boolean _runOnce(@NotNull PropertiesComponent storage,
                                  @NotNull @NonNls String id,
                                  boolean ideAware,
                                  @NotNull Runnable activity) {
    //noinspection SynchronizationOnLocalVariableOrMethodParameter
    synchronized (storage) {
      String key = createKey(id, ideAware);
      if (storage.isTrueValue(key)) {
        return false;
      }
      storage.setValue(key, true);
    }
    activity.run();
    return true;
  }

  private static @NonNls String createKey(@NotNull String id, boolean ideAware) {
    String key = "RunOnceActivity.";
    if (ideAware) {
      key += PlatformUtils.getPlatformPrefix() + ".";
    }
    return  key + id;
  }
}

File: platform/ide-core/src/com/intellij/ide/util/RunOnceUtil.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.ide.util

import com.intellij.openapi.components.serviceAsync
import com.intellij.openapi.project.Project
import org.jetbrains.annotations.ApiStatus.Internal
import org.jetbrains.annotations.NonNls

/**
 * Run a task only once in the history for a project or application.
 *
 * @author Konstantin Bulenkov
 */
object RunOnceUtil {
  /**
   * Perform the task if it was not performed before for the given project.
   *
   * @param id unique ID for the task
   * @return `true` if the task was performed, `false` if the task had already been performed before.
   */
  @JvmStatic
  fun runOnceForProject(project: Project, id: @NonNls String, task: Runnable): Boolean {
    return doRunOnce(storage = PropertiesComponent.getInstance(project), id = id, activity = task)
  }

  /**
   * Perform the task if it was not performed before for this application (running IDE instance).
   *
   * @param id unique ID for the task
   * @return `true` if the task was performed, `false` if the task had already been performed before.
   */
  @JvmStatic
  fun runOnceForApp(id: @NonNls String, task: Runnable): Boolean {
    return doRunOnce(storage = PropertiesComponent.getInstance(), id = id, activity = task)
  }
}

suspend fun runOnceForApp(id: @NonNls String, task: suspend () -> Unit): Boolean {
  val key = createKey(id)
  if (serviceAsync<PropertiesComponent>().updateValue(key, true)) {
    task()
    return true
  }
  else {
    return false
  }
}

@Internal
suspend fun runOnceForProject(project: Project, id: @NonNls String, task: suspend () -> Unit): Boolean {
  val key = createKey(id = id)
  if (project.serviceAsync<PropertiesComponent>().updateValue(key, true)) {
    task()
    return true
  }
  else {
    return false
  }
}

private fun doRunOnce(storage: PropertiesComponent, id: @NonNls String, activity: Runnable): Boolean {
  val key = createKey(id)
  if (!storage.updateValue(key, true)) {
    return false
  }

  storage.setValue(key, true)
  activity.run()
  return true
}

private fun createKey(id: String): @NonNls String = "RunOnceActivity.$id"
